lib/std/math.zig

```diff
@@ -352,6 +352,21 @@ pub fn shlExact(comptime T: type, a: T, shift_amt: Log2Int(T)) !T {
 /// A negative shift amount results in a right shift.
 pub fn shl(comptime T: type, a: T, shift_amt: anytype) T {
     const abs_shift_amt = absCast(shift_amt);
+
+    if (@typeInfo(T) == .Vector) {
+        const C = @typeInfo(T).Vector.child;
+        const len = @typeInfo(T).Vector.len;
+        const casted_shift_amt = if (abs_shift_amt >= @typeInfo(C).Int.bits) return @splat(len, @as(C, 0)) else @intCast(Log2Int(C), abs_shift_amt);
+
+        if (@TypeOf(shift_amt) == comptime_int or @typeInfo(@TypeOf(shift_amt)).Int.is_signed) {
+            if (shift_amt < 0) {
+                return a >> @splat(len, casted_shift_amt);
+            }
+        }
+
+        return a << @splat(len, casted_shift_amt);
+    }
+
     const casted_shift_amt = if (abs_shift_amt >= @typeInfo(T).Int.bits) return 0 else @intCast(Log2Int(T), abs_shift_amt);
```

Comment:
What about something like:

```suggestion
    const casted_shift_amt = blk: {
        if (@typeInfo(T) == .Vector) {
            const C = @typeInfo(T).Vector.child;
            const len = @typeInfo(T).Vector.len;
            if (abs_shift_amt >= @typeInfo(C).Int.bits) return @splat(len, @as(C, 0));
            break :blk @splat(len, @intCast(Log2Int(C), abs_shift_amt));
        } else {
            if (abs_shift_amt >= @typeInfo(T).Int.bits) return 0;
            break :blk @intCast(Log2Int(T), abs_shift_amt);
        }
    };
```

Reply:
What follows would need to be rewritten as well, with another branch. So we end up with something like:

```zig
    const abs_shift_amt = absCast(shift_amt);
    const casted_shift_amt = blk: {
        if (@typeInfo(T) == .Vector) {
            const C = @typeInfo(T).Vector.child;
            if (abs_shift_amt >= @typeInfo(C).Int.bits) {
                return @splat(@typeInfo(T).Vector.len, @as(C, 0));
            } else {
                break :blk @intCast(Log2Int(C), abs_shift_amt);
            }
        } else {
            if (abs_shift_amt >= @typeInfo(T).Int.bits) {
                return 0;
            } else {
                break :blk @intCast(Log2Int(T), abs_shift_amt);
            }
        }
    };
    if (@typeInfo(T) == .Vector) {
        if (@TypeOf(shift_amt) == comptime_int or @typeInfo(@TypeOf(shift_amt)).Int.is_signed) {
            if (shift_amt < 0) {
                return a >> @splat(@typeInfo(T).Vector.len, casted_shift_amt);
            }
        }
        return a << @splat(@typeInfo(T).Vector.len, casted_shift_amt);
    } else {
        if (@TypeOf(shift_amt) == comptime_int or @typeInfo(@TypeOf(shift_amt)).Int.is_signed) {
            if (shift_amt < 0) {
                return a >> casted_shift_amt;
            }
        }
        return a << casted_shift_amt;
    }
```

This is longer and looks more difficult to follow IMHO.

Reply:
Or I probably missed something :) What would the entire function look like?

Reply:
the diff I gave you was the entire change:
```zig
        if (@TypeOf(shift_amt) == comptime_int or @typeInfo(@TypeOf(shift_amt)).Int.is_signed) {
            if (shift_amt < 0) {
                return a >> casted_shift_amt;
            }
        }
        return a << casted_shift_amt;
```
can be shared by the different types: for an integer `casted_shift_amt` is an integer; for a vector, `casted_shift_amt` is a vector

Reply:
Oh, very nice!